### PR TITLE
feat(locksmith) - Generate certification image

### PR DESCRIPTION
**Pull request by kalidiagne**
# Description

Add endpoint to generate certification 

- [x] Add new endpoint to generate certificates
- [x] Add tests 
- [x] Refactor `Certificate` component with inline style 
- [x] Add email attachments 
 
Screenshot 2023-04-11 at 20 30 34

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: locksmith/src/utils/certification.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import { addressMinify } from './strings'
+import satori from 'satori'
+import dayjs from 'dayjs'
+
+import { readFileSync } from 'fs'
+const inter400 = readFileSync('src/fonts/inter-400.woff')
+const inter700 = readFileSync('src/fonts/inter-700.woff')
+
+export const MAX_UINT =
+  '115792089237316195423570985008687907853269984665640564039457584007913129639935'
```
Comment (searchableguy) [inline locksmith/src/utils/certification.tsx:16]: move it to a separate file? I thought this was already a constant.

File: locksmith/src/utils/certification.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        width: '100vw',
+      }}
+    >
+      <Certificate
+        name={metadata?.name || ''}
+        description={metadata.description}
+        tokenId={metadata?.tokenId}
+        owner={metadata?.owner}
+        issuer={object?.certification_issuer}
+        image={metadata?.image}
```
Comment (searchableguy) [inline locksmith/src/utils/certification.tsx:76]: satori earlier had some issues with fetching mime data. Is this working properly now after update? Can you try with a few different image/metadata?

File: locksmith/src/utils/certification.tsx (hunk 1 above)
Comment (kalidiagne) [reply locksmith/src/utils/certification.tsx:16]: not in locksmith, only in `unlock-app`, maybe we can add a shared lib later

File: locksmith/src/utils/strings.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export function addressMinify(address: string): string {
+  return `${address?.substring(0, 6)}...${address?.substring(
+    address.length - 4
+  )}`
+}
```
Comment (searchableguy) [inline locksmith/src/utils/strings.ts:5]: remove. Import from ui library 'minifyAddress'

Comment (kalidiagne) [reply locksmith/src/utils/strings.ts:5]: oh awesome, importing from there :)

File: locksmith/constants.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export const MAX_UINT =
```
Comment (julien51) [inline locksmith/constants.ts:1]: Do we not have this already? If not please use `ethers` which already defines it.

File: locksmith/src/operations/wedlocksOperations.ts [hunk 5]
```diff
@@ -214,6 +215,20 @@ const getAttachments = async ({
...
+  if (isCertification && network && tokenId) {
+    const certificate = await createCertificate({
+      network,
+      lockAddress,
+      tokenId,
+    })
+    const svg = new resvg.Resvg(certificate)
+    const pngData = svg.render()
+    const pngBuffer = pngData.asPng()
+    const dataURI = `data:image/png;base64,${pngBuffer.toString('base64')}`
```
Comment (julien51) [inline locksmith/src/operations/wedlocksOperations.ts:228]: Why do you need to render as png and not use svg since you are doing a data uri anyway?

File: locksmith/src/utils/certification.tsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        description={metadata.description}
+        tokenId={metadata?.tokenId}
+        owner={metadata?.owner}
+        issuer={object?.certification_issuer}
+        image={imageBase64}
+        lockAddress={minifyAddress(lockAddress)}
+        network={network}
+        networkName={networks[network]?.name}
+        // can't use minifyAddress because is checking if the address is valid
+        transactionsHash={`${transactionHash.slice(
```
Comment (julien51) [inline locksmith/src/utils/certification.tsx:89]: Please create a util function and add to the core package?

File: locksmith/src/utils/image.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+export const imageToBase64 = async (url: string) => {
+  const response = await fetch(url, {
```
Comment (julien51) [inline locksmith/src/utils/image.ts:2]: Why are we fetching the image? It wastes resources because this image is already stored here, Let's just retrieve it.

File: packages/ui/lib/components/Certificate/Certificate.tsx [hunk 8]
```diff
@@ -18,11 +17,12 @@ interface CertificateProps {
   badge?: string
   transactionsHash?: ReactNode
   externalUrl?: string
+  isMobile?: boolean
 }
 
 interface Props {
   children?: ReactNode
-  size?: Size
+  size?: string
```
Comment (julien51) [inline packages/ui/lib/components/Certificate/Certificate.tsx:25]: Why the change?

File: locksmith/src/utils/certification.tsx (hunk 6 above)
Comment (kalidiagne) [reply locksmith/src/utils/certification.tsx:89]: we have it already in `ui`:

```
export const minifyAddress = (address: string) => {
  const checked = ethers.utils.isAddress(address)
  return checked
    ? `${address.slice(0, 5)}...${address.slice(address.length - 5)}`
    : address
}
```
should not work for transaction hash because we verify that we pass an address for minify

File: locksmith/constants.ts (hunk 4 above)
Comment (kalidiagne) [reply locksmith/constants.ts:1]: not present in `locksmith`, replacing with `ethers`

File: locksmith/src/operations/wedlocksOperations.ts (hunk 5 above)
Comment (kalidiagne) [reply locksmith/src/operations/wedlocksOperations.ts]: refactored, but still i think we need to keep it as is, the SVG need to be decoded otherwise

i think @searchableguy  can give more details about the implementation here

File: packages/ui/lib/components/Certificate/Certificate.tsx (hunk 8 above)
Comment (kalidiagne) [reply packages/ui/lib/components/Certificate/Certificate.tsx]: ah! restored

File: locksmith/src/operations/wedlocksOperations.ts (hunk 5 above)
Comment (searchableguy) [reply locksmith/src/operations/wedlocksOperations.ts]: I'm confused. If you are talking about the reason why I convert svg to png for tickets, that's for displaying the ticket in emails and make it easy to print.

File: locksmith/src/utils/certification.tsx (hunk 6 above)
Comment (julien51) [reply locksmith/src/utils/certification.tsx:90]: ok but there is no value in displaying a truncated version if there is no link, because nobody can do anything useful with a truncated hash... so either you show the full hash or you point to the block explorer.

Comment (kalidiagne) [reply locksmith/src/utils/certification.tsx:90]: In this case for the generated image the link is not present, but is present on the certification detail https://github.com/unlock-protocol/unlock/blob/master/unlock-app/src/components/content/certification/CertificationDetails.tsx#L215
